### ISSUE TITLE
TileSet Collision Editor - buggy behaviour when releasing key combinations in different order

### ISSUE THREAD

**Issue by d0mk** [godotengine/godot#54481]
### Godot version

3.3.4.stable.mono.official.faf3f883d

### System information

Windows 10

### Issue description

I'm following a YT tutorial on making a 2D game in Godot, and currently I'm adding collision rectangles to my TileSet in the TileSet editor, Collision tab.

I've noticed that the "New Rectangle" button can be activated using "Shift+R" shortcut, so I started using it to make the process faster, but then I've encountered some bugs:
- if the selection tool (mouse pointer icon) is active and "Shift+R" is pressed too fast, both icons stay highlighted and clicking on the tile grid has no effect (it acts as if only the selection tool was active)
- if the "New Rectangle" tool is active and "Shift+P" (New Polygon) is clicked too fast, both icons stay highlighted and clicking on the grid has the effect of "New Rectangle" tool
- if the "New Polygon" tool is active and "Shift+R" is clicked too fast, both icons stay highlighted and clicking on the grid has the effect of "New Polygon" tool

Actually, I think "clicking too fast" just means releasing the keys in different order, for example: if I have the selection tool active and I press Shift, then press R, then release R, then release Shift, the tool is properly changed to "New Rectangle". But if I press Shift, then press R, then release Shift, then release R, the bug appears. So it seems like releasing the keys in certain order matters, that's why doing it too fast may be causing issues.

### Steps to reproduce

1. Open the TileSet editor in the attached project.
2. Navigate to Collision tab.
3. Select the "Selection" tool.
4. Click on any cell on the grid.
5. Press and hold left Shift.
6. Press and hold R.
7. Release left Shift.
8. Release R.

After these steps, "Selection" and "New Rectangle" icons stay highlighted, clicking on the grid has the effect of "Selection" tool. Same stays true if you try "Shift+P" or start on different tool than "Selection".

### Minimal reproduction project

[minimal-example.zip](https://github.com/godotengine/godot/files/7453449/minimal-example.zip)

**Comment by EricEzaM** [comment]
Duplicate of #45033 I think. You did a good job identifying that the order matters - I believe that is the correct diagnosis. Technically the issues are different but the underlying problem is the same between the two.

**Comment by PikaBlue107** [comment]
Seconding this issue. Note that this behavior **also occurs with the "Delete Selected Shape" shortcut** (Shift+BackSpace).

Demo of bug attached, using [AutoHotKey's OSAK](https://www.autohotkey.com/docs/scripts/index.htm#KeyboardOnScreen). This script unfortunately doesn't show the sequence "Shift down, R down, Shift up, R up" explicitly, instead highlighting Shift until R is released as well.

https://user-images.githubusercontent.com/6174468/155889014-f3393851-4cf2-48a3-a43e-8528e5ee24fc.mp4

**Comment by coffe789** [comment]
I've done some investigating on this issue and I've found it isn't exclusive to the tileset editor, and it applies to every button with ACTION_MODE_BUTTON_RELEASE. If you press the shortcut key for a given button, then press either shift, alt, meta, or ctrl, and then let go of the buttons (letting go of the shortcut key last), letting go of the shortcut is not detected. This explains why the button stays highlighted but doesn't get pressed. 
Should also apply to #45033

**Comment by PikaBlue107** [comment]
**Can anyone check if this occurs on other OS's?**

I'm also on Windows and trying to trace this one (new contributor here). I've identified that the "wrong" shortcut produces flawed `InputEventKey`s:
(Key combo retrieved from `event.as_text()`, "Down"/"Up" retrieved from `event.pressed`)

| Keypress | InputEventKey  |
|-----------|------------------|
| ↓Shift      | Shift Down        |
| ↓R           |  Shift+R Down  |
| ↑Shift      | Shift Up            |
| ↑R           | R Up                 |

whereas the "correct" shortcut produces correct ones:

| Keypress | InputEventKey  |
|-----------|------------------|
| ↓Shift      | Shift Down        |
| ↓R           |  Shift+R Down  |
| ↑R           | Shift+R Up (trigger shortcut)       |
| ↑Shift      | Shift Up                 |

I'm currently trying to trace where in the `3.x` codebase this `InputEventKey` is being created. This behavior shows up in `BaseButton` and `InputDefault`. The next layer seems to be `OS_Windows`, but that seems to imply that this is windows-specific. Hence - can anyone try this bug on another OS?

**Comment by Zireael07** [comment]
Yes, I'm 100% sure this is editor-wide, not just tileset collision editor. Can't find the other report atm.

**Comment by jhaakma** [comment]
This is particularly annoying in the TileSet editor because of how cumbersome it is to add collision to individual tiles.

I've confirmed the issue exists on both Windows 10 and on MacOS.

**Comment by coffe789** [comment]
I can confirm this also occurs on Linux (tested using Debian 11)
